java/arcs/core/util/Scheduler.kt

```diff
@@ -0,0 +1,242 @@
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.util
+
+import kotlin.coroutines.CoroutineContext
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.update
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.TimeoutCancellationException
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.withTimeout
+
+/**
+ * The [Scheduler] is responsible for scheduling the execution of a batch of [Task]s (known as an
+ * [Agenda]), at a defined maximum [scheduleRateHz], where each agenda is limited to running within
+ * the specified [agendaProcessingTimeout] window.
+ */
+class Scheduler(
+    private val time: Time,
+    context: CoroutineContext,
+    private val agendaProcessingTimeout: Long = DEFAULT_AGENDA_PROCESSING_TIMEOUT,
+    private val scheduleRateHz: Int = DEFAULT_SCHEDULE_RATE_HZ
+) {
+    private val log = TaggedLog { "Scheduler(${hashCode()})" }
+    /* internal */
+    val launches = atomic(0)
+    /* internal */
+    val loops = atomic(0)
+    private var processingJob: Job? = null
+    private val scope = CoroutineScope(context)
+    private val isIdle = atomic(true)
+    private val isPaused = atomic(false)
+    private val agenda = atomic(Agenda())
+
+    /** Schedule a single [Task] to be run as part of the next agenda. */
+    fun schedule(task: Task) {
+        agenda.update { it.addTask(task) }
+        if (isIdle.value) scope.startProcessingJob()
+    }
+
+    /** Schedule some [Task]s to be run as part of the next agenda. */
+    fun schedule(tasks: Iterable<Task>) {
+        agenda.update { it.addTasks(tasks) }
+        if (isIdle.value) scope.startProcessingJob()
+    }
+
+    /** Pause evaluation of the agenda. */
+    fun pause() {
+        isPaused.value = true
+    }
+
+    /** Resume evaluation of scheduled [Task]s in the agenda. */
+    fun resume() {
+        if (!isPaused.compareAndSet(expect = true, update = false)) return
+        // If we were paused, we can re-start.
+        scope.startProcessingJob()
+    }
+
+    /** Wait for the [Scheduler] to become idle. */
+    suspend fun waitForIdle() {
+        processingJob?.join()
+    }
+
+    private fun CoroutineScope.startProcessingJob() {
+        isIdle.value = false
+        processingJob = launch {
+            launches.incrementAndGet()
+
+            var shallContinue = true
+            while (shallContinue) {
+                val startTime = time.currentTimeMillis
+                shallContinue = withTimeout(agendaProcessingTimeout) { process() }
+                val elapsed = time.currentTimeMillis - startTime
+
+                if (shallContinue) {
+                    loops.incrementAndGet()
+
+                    // If the operation was super fast, let's delay a little bit to allow the
+                    // thread to chill out.
```

Comment:
I'm curious, what's the purpose behind this? Why not just run through all the tasks as quickly as possible?

Reply:
I think it is a holdover from Typescript only having one thread, so you have to yield or the UI locks up?

It still might be useful as a mitigation against denial-of-service to rate limit the speed at which particles can operate on Handles, however, I feel this may complicate Ian's benchmarking results.

Reply:
Yeah, the primary reason I wanted to add this was to avoid a situation where an infinite loop of mutations from handles in the same arc (thus sharing this scheduler) would result in the scheduler thread itself taking up 100% of CPU time.